File: SudokuSolver_Java_No_Streams/src/main/sudokusolver/javanostreams/SudokuSolver.java
package sudokusolver.javanostreams;

import sudokusolver.javanostreams.logic.BruteForce;
import sudokusolver.javanostreams.logic.MultipleSolutionsException;
import sudokusolver.javanostreams.logic.NoSolutionsException;
import sudokusolver.javanostreams.logic.diabolical.AlignedPairExclusion;
import sudokusolver.javanostreams.logic.diabolical.BUG;
import sudokusolver.javanostreams.logic.diabolical.ExtendedUniqueRectangles;
import sudokusolver.javanostreams.logic.diabolical.HiddenUniqueRectangles;
import sudokusolver.javanostreams.logic.diabolical.Jellyfish;
import sudokusolver.javanostreams.logic.diabolical.Medusa;
import sudokusolver.javanostreams.logic.diabolical.UniqueRectangles;
import sudokusolver.javanostreams.logic.diabolical.WXYZWing;
import sudokusolver.javanostreams.logic.diabolical.XCycles;
import sudokusolver.javanostreams.logic.diabolical.XYChains;
import sudokusolver.javanostreams.logic.extreme.AlternatingInferenceChains;
import sudokusolver.javanostreams.logic.extreme.EmptyRectangles;
import sudokusolver.javanostreams.logic.extreme.FinnedSwordfish;
import sudokusolver.javanostreams.logic.extreme.FinnedXWing;
import sudokusolver.javanostreams.logic.extreme.GroupedXCycles;
import sudokusolver.javanostreams.logic.extreme.SueDeCoq;
import sudokusolver.javanostreams.logic.simple.BoxLineReduction;
import sudokusolver.javanostreams.logic.simple.HiddenPairs;
import sudokusolver.javanostreams.logic.simple.HiddenQuads;
import sudokusolver.javanostreams.logic.simple.HiddenSingles;
import sudokusolver.javanostreams.logic.simple.HiddenTriples;
import sudokusolver.javanostreams.logic.simple.NakedPairs;
import sudokusolver.javanostreams.logic.simple.NakedQuads;
import sudokusolver.javanostreams.logic.simple.NakedSingles;
import sudokusolver.javanostreams.logic.simple.NakedTriples;
import sudokusolver.javanostreams.logic.simple.PointingPairsPointingTriples;
import sudokusolver.javanostreams.logic.simple.PruneCandidates;
import sudokusolver.javanostreams.logic.tough.SimpleColoring;
import sudokusolver.javanostreams.logic.tough.Swordfish;
import sudokusolver.javanostreams.logic.tough.XWing;
import sudokusolver.javanostreams.logic.tough.XYZWing;
import sudokusolver.javanostreams.logic.tough.YWing;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;

public class SudokuSolver {
    private static final List<Function<Board<Cell>, List<? extends BoardModification>>> SOLUTIONS = List.of(
            //Start of simple solutions.
            PruneCandidates::pruneCandidates,
            NakedSingles::nakedSingles,
            HiddenSingles::hiddenSingles,
            NakedPairs::nakedPairs,
            NakedTriples::nakedTriples,
            HiddenPairs::hiddenPairs,
            HiddenTriples::hiddenTriples,
            NakedQuads::nakedQuads,
            HiddenQuads::hiddenQuads,
            PointingPairsPointingTriples::pointingPairsPointingTriples,
            BoxLineReduction::boxLineReduction,
            //Start of tough solutions.
            XWing::xWing,
            SimpleColoring::simpleColoringRule2,
            SimpleColoring::simpleColoringRule4,
            YWing::yWing,
            Swordfish::swordfish,
            XYZWing::xyzWing,
            //Start of diabolical solutions.
            XCycles::xCyclesRule1,
            XCycles::xCyclesRule2,
            XCycles::xCyclesRule3,
            b -> BUG.bug(b).map(List::of).orElseGet(Collections::emptyList),
            XYChains::xyChains,
            Medusa::medusaRule1,
            Medusa::medusaRule2,
            Medusa::medusaRule3,
            Medusa::medusaRule4,
            Medusa::medusaRule5,
            Medusa::medusaRule6,
            Jellyfish::jellyfish,
            UniqueRectangles::uniqueRectanglesType1,
            UniqueRectangles::uniqueRectanglesType2,
            UniqueRectangles::uniqueRectanglesType3,
            UniqueRectangles::uniqueRectanglesType3BWithTriplePseudoCells,
            UniqueRectangles::uniqueRectanglesType4,
            UniqueRectangles::uniqueRectanglesType5,
            ExtendedUniqueRectangles::extendedUniqueRectangles,
            HiddenUniqueRectangles::hiddenUniqueRectangles,
            WXYZWing::wxyzWing,
            AlignedPairExclusion::alignedPairExclusion,
            //Start of extreme solutions.
            GroupedXCycles::groupedXCyclesRule1,
            GroupedXCycles::groupedXCyclesRule2,
            GroupedXCycles::groupedXCyclesRule3,
            EmptyRectangles::emptyRectangles,
            FinnedXWing::finnedXWing,
            FinnedSwordfish::finnedSwordfish,
            AlternatingInferenceChains::alternatingInferenceChainsRule1,
            AlternatingInferenceChains::alternatingInferenceChainsRule2,
            AlternatingInferenceChains::alternatingInferenceChainsRule3,
            SueDeCoq::sueDeCoq
    );

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("usage: java sudokusolver.javanostreams.SudokuSolver board");
        } else {
            var board = args[0];
            var showError = board.length() != Board.UNIT_SIZE_SQUARED;
            for (int i = 0; !showError && i < board.length(); i++) {
                if (board.charAt(i) < '0' || board.charAt(i) > '9') {
                    showError = true;
                    break;
                }
            }
            if (showError) {
                System.out.println("board must be " + Board.UNIT_SIZE_SQUARED + " numbers with blanks expressed as 0");
            } else {
                try {
                    System.out.println(solve(BoardFactory.parseOptionalBoard(board)));
                } catch (NoSolutionsException e) {
                    System.out.println("No Solutions");
                } catch (MultipleSolutionsException e) {
                    System.out.println("Multiple Solutions");
                } catch (UnableToSolveException e) {
                    System.out.println(e.getMessage());
                }
            }
        }
    }

    public static Board<SudokuNumber> solve(
            Board<Optional<SudokuNumber>> input
    ) throws NoSolutionsException, MultipleSolutionsException, UnableToSolveException {
        var bruteForceSolution = BruteForce.bruteForce(input);
        var board = BoardFactory.createCellBoard(input);
        List<? extends BoardModification> modifications;
        do {
            var allSolved = true;
            for (var cell : board.getCells()) {
                if (cell instanceof UnsolvedCell) {
                    allSolved = false;
                    break;
                }
            }
            if (allSolved) {
                return bruteForceSolution;
            }
            modifications = performNextSolution(board);
            for (var modification : modifications) {
                var row = modification.row();
                var column = modification.column();
                var cell = board.get(row, column);
                if (cell instanceof UnsolvedCell unsolved) {
                    var knownSolution = bruteForceSolution.get(row, column);
                    if (modification instanceof RemoveCandidates removeCandidates) {
                        for (var candidate : removeCandidates.candidates()) {
                            if (candidate == knownSolution) {
                                var message = "Cannot remove candidate " + candidate + " from [" + row + ", " + column +
                                        ']';
                                throw new IllegalStateException(message);
                            }
                            if (!unsolved.candidates().contains(candidate)) {
                                var message = candidate + " is not a candidate of [" + row + ", " + column + ']';
                                throw new IllegalStateException(message);
                            }
                        }
                        unsolved.candidates().removeAll(removeCandidates.candidates());
                    } else if (modification instanceof SetValue setValue) {
                        var value = setValue.value();
                        if (value != knownSolution) {
                            var message = "Cannot set value " + value + " to [" + row + ", " + column +
                                    "]. Solution is " + knownSolution;
                            throw new IllegalStateException(message);
                        }
                        board.set(row, column, new SolvedCell(row, column, value));
                    }
                } else if (cell instanceof SolvedCell) {
                    throw new IllegalStateException("[" + row + ", " + column + "] is already solved.");
                }
            }
        } while (!modifications.isEmpty());
        throw new UnableToSolveException(board);
    }

    private static List<? extends BoardModification> performNextSolution(Board<Cell> board) {
        for (var solution : SOLUTIONS) {
            var modifications = solution.apply(board);
            if (!modifications.isEmpty()) {
                return modifications;
            }
        }
        return Collections.emptyList();
    }
}

File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/SudokuSolver.kt
package sudokusolver.kotlin

import sudokusolver.kotlin.logic.MultipleSolutions
import sudokusolver.kotlin.logic.NoSolutions
import sudokusolver.kotlin.logic.SingleSolution
import sudokusolver.kotlin.logic.bruteForce
import sudokusolver.kotlin.logic.diabolical.alignedPairExclusion
import sudokusolver.kotlin.logic.diabolical.bug
import sudokusolver.kotlin.logic.diabolical.extendedUniqueRectangles
import sudokusolver.kotlin.logic.diabolical.hiddenUniqueRectangles
import sudokusolver.kotlin.logic.diabolical.jellyfish
import sudokusolver.kotlin.logic.diabolical.medusaRule1
import sudokusolver.kotlin.logic.diabolical.medusaRule2
import sudokusolver.kotlin.logic.diabolical.medusaRule3
import sudokusolver.kotlin.logic.diabolical.medusaRule4
import sudokusolver.kotlin.logic.diabolical.medusaRule5
import sudokusolver.kotlin.logic.diabolical.medusaRule6
import sudokusolver.kotlin.logic.diabolical.uniqueRectanglesType1
import sudokusolver.kotlin.logic.diabolical.uniqueRectanglesType2
import sudokusolver.kotlin.logic.diabolical.uniqueRectanglesType3
import sudokusolver.kotlin.logic.diabolical.uniqueRectanglesType3BWithTriplePseudoCells
import sudokusolver.kotlin.logic.diabolical.uniqueRectanglesType4
import sudokusolver.kotlin.logic.diabolical.uniqueRectanglesType5
import sudokusolver.kotlin.logic.diabolical.wxyzWing
import sudokusolver.kotlin.logic.diabolical.xCyclesRule1
import sudokusolver.kotlin.logic.diabolical.xCyclesRule2
import sudokusolver.kotlin.logic.diabolical.xCyclesRule3
import sudokusolver.kotlin.logic.diabolical.xyChains
import sudokusolver.kotlin.logic.extreme.emptyRectangles
import sudokusolver.kotlin.logic.extreme.finnedSwordfish
import sudokusolver.kotlin.logic.extreme.finnedXWing
import sudokusolver.kotlin.logic.extreme.groupedXCyclesRule1
import sudokusolver.kotlin.logic.extreme.groupedXCyclesRule2
import sudokusolver.kotlin.logic.extreme.groupedXCyclesRule3
import sudokusolver.kotlin.logic.simple.boxLineReduction
import sudokusolver.kotlin.logic.simple.hiddenPairs
import sudokusolver.kotlin.logic.simple.hiddenQuads
import sudokusolver.kotlin.logic.simple.hiddenSingles
import sudokusolver.kotlin.logic.simple.hiddenTriples
import sudokusolver.kotlin.logic.simple.nakedPairs
import sudokusolver.kotlin.logic.simple.nakedQuads
import sudokusolver.kotlin.logic.simple.nakedSingles
import sudokusolver.kotlin.logic.simple.nakedTriples
import sudokusolver.kotlin.logic.simple.pointingPairsPointingTriples
import sudokusolver.kotlin.logic.simple.pruneCandidates
import sudokusolver.kotlin.logic.tough.simpleColoringRule2
import sudokusolver.kotlin.logic.tough.simpleColoringRule4
import sudokusolver.kotlin.logic.tough.swordfish
import sudokusolver.kotlin.logic.tough.xWing
import sudokusolver.kotlin.logic.tough.xyzWing
import sudokusolver.kotlin.logic.tough.yWing

fun main(args: Array<String>) {
    if (args.size != 1) {
        println("usage: SudokuSolverKt board")
    } else {
        val board = args.single()
        if (board.length != UNIT_SIZE_SQUARED || board.any { it !in '0'..'9' }) {
            println("board must be $UNIT_SIZE_SQUARED numbers with blanks expressed as 0")
        } else {
            println(
                when (val solution = solve(parseOptionalBoard(board))) {
                    InvalidNoSolutions -> "No Solutions"
                    InvalidMultipleSolutions -> "Multiple Solutions"
                    is Solution -> solution.board

                    is UnableToSolve -> {
                        """
                            |Unable to solve:
                            |${solution.board.toString().lines().joinToString("\n")}
                            |
                            |Simple String: ${solution.board.toSimpleString()}
                            |
                            |With Candidates:
                            |${solution.board.toStringWithCandidates()}
                        """.trimMargin()
                    }
                }
            )
        }
    }
}

private sealed interface SolveResult
private object InvalidNoSolutions : SolveResult
private object InvalidMultipleSolutions : SolveResult
private class Solution(val board: Board<SudokuNumber>) : SolveResult
private class UnableToSolve(val board: Board<Cell>) : SolveResult

private fun solve(input: Board<SudokuNumber?>): SolveResult {
    when (val bruteForceSolution = bruteForce(input)) {
        NoSolutions -> return InvalidNoSolutions
        MultipleSolutions -> return InvalidMultipleSolutions

        is SingleSolution -> {
            val mutableBoard = createMutableCellBoard(input)
            do {
                if (mutableBoard.cells.filterIsInstance<UnsolvedCell>().isEmpty()) {
                    return Solution(bruteForceSolution.solution)
                }
                /*
                 * Why do I convert the MutableBoard to an immutable Board just to pass the immutable one to the logic
                 * functions? I wanted the logic functions to have the guarantee that the board will not change while
                 * they are running. Even if I have the logic functions take an AbstractBoard, that would be no
                 * guarantee. That would simply prevent the logic function from mutating the board without casting it.
                 * The only way to have mutability here and true immutability in the logic functions is to copy the
                 * contents from MutableBoard to Board.
                 *
                 * In this case, Rust would be better here than the JVM. In Rust, the board would be owned and mutable
                 * here. Read-only references would then be passed to the logic functions. The Rust compiler would
                 * ensure that the board is never modified here while read-only references are borrowed by the logic
                 * functions. With Rust, we could have mutability here, guaranteed immutability in the logic functions,
                 * and no copying.
                 */
                val board = mutableBoard.toBoard()
                //Start of simple solutions.
                val modifications = pruneCandidates(board)
                    .ifEmpty { nakedSingles(board) }
                    .ifEmpty { hiddenSingles(board) }
                    .ifEmpty { nakedPairs(board) }
                    .ifEmpty { nakedTriples(board) }
                    .ifEmpty { hiddenPairs(board) }
                    .ifEmpty { hiddenTriples(board) }
                    .ifEmpty { nakedQuads(board) }
                    .ifEmpty { hiddenQuads(board) }
                    .ifEmpty { pointingPairsPointingTriples(board) }
                    .ifEmpty { boxLineReduction(board) }
                    //Start of tough solutions.
                    .ifEmpty { xWing(board) }
                    .ifEmpty { simpleColoringRule2(board) }
                    .ifEmpty { simpleColoringRule4(board) }
                    .ifEmpty { yWing(board) }
                    .ifEmpty { swordfish(board) }
                    .ifEmpty { xyzWing(board) }
                    //Start of diabolical solutions.
                    .ifEmpty { xCyclesRule1(board) }
                    .ifEmpty { xCyclesRule2(board) }
                    .ifEmpty { xCyclesRule3(board) }
                    .ifEmpty { listOfNotNull(bug(board)) }
                    .ifEmpty { xyChains(board) }
                    .ifEmpty { medusaRule1(board) }
                    .ifEmpty { medusaRule2(board) }
                    .ifEmpty { medusaRule3(board) }
                    .ifEmpty { medusaRule4(board) }
                    .ifEmpty { medusaRule5(board) }
                    .ifEmpty { medusaRule6(board) }
                    .ifEmpty { jellyfish(board) }
                    .ifEmpty { uniqueRectanglesType1(board) }
                    .ifEmpty { uniqueRectanglesType2(board) }
                    .ifEmpty { uniqueRectanglesType3(board) }
                    .ifEmpty { uniqueRectanglesType3BWithTriplePseudoCells(board) }
                    .ifEmpty { uniqueRectanglesType4(board) }
                    .ifEmpty { uniqueRectanglesType5(board) }
                    .ifEmpty { extendedUniqueRectangles(board) }
                    .ifEmpty { hiddenUniqueRectangles(board) }
                    .ifEmpty { wxyzWing(board) }
                    .ifEmpty { alignedPairExclusion(board) }
                    //Start of extreme solutions.
                    .ifEmpty { groupedXCyclesRule1(board) }
                    .ifEmpty { groupedXCyclesRule2(board) }
                    .ifEmpty { groupedXCyclesRule3(board) }
                    .ifEmpty { emptyRectangles(board) }
                    .ifEmpty { finnedXWing(board) }
                    .ifEmpty { finnedSwordfish(board) }
                modifications.forEach { modification ->
                    val row = modification.row
                    val column = modification.column
                    val cell = mutableBoard[row, column]
                    check(cell is UnsolvedCell) { "[$row, $column] is already solved." }
                    val knownSolution = bruteForceSolution.solution[row, column]
                    when (modification) {
                        is RemoveCandidates -> {
                            modification.candidates.forEach { candidate ->
                                check(candidate != knownSolution) {
                                    "Cannot remove candidate $candidate from [$row, $column]"
                                }
                                check(candidate in cell.candidates) {
                                    "$candidate is not a candidate of [$row, $column]"
                                }
                                cell.candidates -= candidate
                            }
                        }

                        is SetValue -> {
                            val value = modification.value
                            check(value == knownSolution) {
                                "Cannot set value $value to [$row, $column]. Solution is $knownSolution"
                            }
                            mutableBoard[row, column] = SolvedCell(row, column, value)
                        }
                    }
                }
            } while (modifications.isNotEmpty())

            return UnableToSolve(mutableBoard.toBoard())
        }
    }
}